leshan-tl-javacoap-server/src/main/java/org/eclipse/leshan/transport/javacoap/endpoint/JavaCoapServerEndpoint.java

```diff
@@ -0,0 +1,233 @@
+import org.eclipse.leshan.core.request.ObserveRequest;
+import org.eclipse.leshan.core.request.exception.SendFailedException;
+import org.eclipse.leshan.core.response.ErrorCallback;
+import org.eclipse.leshan.core.response.LwM2mResponse;
+import org.eclipse.leshan.core.response.ObserveResponse;
+import org.eclipse.leshan.core.response.ResponseCallback;
+import org.eclipse.leshan.server.endpoint.LwM2mServerEndpoint;
+import org.eclipse.leshan.server.endpoint.ServerEndpointToolbox;
+import org.eclipse.leshan.server.observation.LwM2mNotificationReceiver;
+import org.eclipse.leshan.server.profile.ClientProfile;
+import org.eclipse.leshan.server.registration.RegistrationStore;
+import org.eclipse.leshan.server.request.LowerLayerConfig;
+import org.eclipse.leshan.transport.javacoap.observation.ObservationUtil;
+
+import com.mbed.coap.client.CoapClient;
+import com.mbed.coap.client.CoapClientBuilder;
+import com.mbed.coap.client.ObservationConsumer;
+import com.mbed.coap.packet.CoapRequest;
+import com.mbed.coap.packet.CoapResponse;
+import com.mbed.coap.packet.Opaque;
+import com.mbed.coap.server.CoapServer;
+
+public class JavaCoapServerEndpoint implements LwM2mServerEndpoint {
+
+    private final URI endpointUri;
+    private final CoapServer coapServer;
+    private final ServerCoapMessageTranslator translator;
+    private final ServerEndpointToolbox toolbox;
+    private final LwM2mNotificationReceiver notificationReceiver;
+    private final RegistrationStore registrationStore;
+
+    public JavaCoapServerEndpoint(URI endpointUri, CoapServer coapServer, ServerCoapMessageTranslator translator,
+            ServerEndpointToolbox toolbox, LwM2mNotificationReceiver notificationReceiver,
+            RegistrationStore registrationStore) {
+        this.endpointUri = endpointUri;
+        this.coapServer = coapServer;
+        this.translator = translator;
+        this.toolbox = toolbox;
+        this.notificationReceiver = notificationReceiver;
+        this.registrationStore = registrationStore;
+
+    }
+
+    @Override
+    public Protocol getProtocol() {
+        return Protocol.COAP;
+    }
+
+    @Override
+    public URI getURI() {
+        return endpointUri;
+    }
+
+    @Override
+    public <T extends LwM2mResponse> T send(ClientProfile destination, DownlinkRequest<T> request,
+            LowerLayerConfig lowerLayerConfig, long timeoutInMs) throws InterruptedException {
+
+        // Send LWM2M Request
+        CompletableFuture<T> lwm2mResponseFuture = sendLwM2mRequest(destination, request, lowerLayerConfig);
+
+        // Wait synchronously for LWM2M response
+        try {
+            return lwm2mResponseFuture.get(timeoutInMs, TimeUnit.MILLISECONDS);
+        } catch (CompletionException | ExecutionException exception) {
+            // TODO we probably need to enhance this (better translate java-coap exceptions to leshan ones)
+            throw new SendFailedException("Unable to send request", exception);
```

Comment:
At least you could do:
```suggestion
            throw new SendFailedException("Unable to send request", exception.getCause());
```

Reply:
Thx I will do it when I will polish exception handling and timeout :pray: